typescript/sdk/src/ism/metadata/multisig.ts

```diff
@@ -1,39 +1,217 @@
 import { joinSignature, splitSignature } from 'ethers/lib/utils.js';
 
+import { MerkleTreeHook__factory } from '@hyperlane-xyz/core';
 import {
+  Address,
   Checkpoint,
   MerkleProof,
+  S3CheckpointWithId,
   SignatureLike,
+  WithAddress,
   assert,
+  bytes32ToAddress,
   chunk,
   ensure0x,
+  eqAddress,
+  eqAddressEvm,
   fromHexString,
+  rootLogger,
   strip0x,
   toHexString,
 } from '@hyperlane-xyz/utils';
 
-import { ModuleType } from '../types.js';
+import { S3Validator } from '../../aws/validator.js';
+import { HyperlaneCore } from '../../core/HyperlaneCore.js';
+import { MerkleTreeHookConfig } from '../../hook/types.js';
+import { ChainName } from '../../types.js';
+import { IsmType, MultisigIsmConfig } from '../types.js';
+
+import { MetadataBuilder, MetadataContext } from './builder.js';
 
 interface MessageIdMultisigMetadata {
-  type: ModuleType.MESSAGE_ID_MULTISIG;
+  type: IsmType.MESSAGE_ID_MULTISIG;
   signatures: SignatureLike[];
   checkpoint: Omit<Checkpoint, 'mailbox_domain'>;
 }
 
 interface MerkleRootMultisigMetadata
   extends Omit<MessageIdMultisigMetadata, 'type'> {
-  type: ModuleType.MERKLE_ROOT_MULTISIG;
+  type: IsmType.MERKLE_ROOT_MULTISIG;
   proof: MerkleProof;
 }
 
+const MerkleTreeInterface = MerkleTreeHook__factory.createInterface();
+
 const SIGNATURE_LENGTH = 65;
 
 export type MultisigMetadata =
   | MessageIdMultisigMetadata
   | MerkleRootMultisigMetadata;
 
-export class MultisigMetadataBuilder {
-  static encodeSimplePrefix(metadata: MessageIdMultisigMetadata): string {
+export class MultisigMetadataBuilder implements MetadataBuilder {
+  private validatorCache: Record<ChainName, Record<string, S3Validator>> = {};
+
+  constructor(
+    protected readonly core: HyperlaneCore,
+    protected readonly logger = rootLogger.child({
+      module: 'MultisigMetadataBuilder',
+    }),
+  ) {}
+
+  private async s3Validators(
```

Comment:
IMO there's pretty much no reason any of the members + methods in these classes should be private instead of protected. Would be short-sighted